### ISSUE TITLE
Cannot run `npm install` on a newly-cloned repository

### ISSUE THREAD

**Issue by schmod** [angular/angular.js#13123]
Running `npm install` on a freshly-cloned repository fails, and leaves no `node_modules` folder intact:

```
> angularjs@ preinstall /home/schmod/source/angular.js
> node scripts/npm/check-node-modules.js --purge

:-( npm dependencies are stale or in an unknown state!
    Purging 'node_modules'...

> angularjs@ postinstall /home/schmod/source/angular.js
> node scripts/npm/copy-npm-shrinkwrap.js

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
Failed to copy `npm-shrinkwrap.json` to `node_modules/npm-shrinkwrap.cached.json`:
{ [Error: ENOENT: no such file or directory, open 'node_modules/npm-shrinkwrap.cached.json']
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: 'node_modules/npm-shrinkwrap.cached.json' }
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

NPM v3.3.8; Node 4.2.1; Ubuntu 15.04

**Comment by schmod** [comment]
It looks like the changes introduced in #12792 / d3de0066b0db40b09243c3845fe810401cf3c7ef will delete the user's `node_modules` folder when the `preinstall` script runs if a NPM Shrinkwrap cache file is not present.

I'm assuming that NPM v3.x doesn't add this file before the `preinstall` scripts are run, which is causing the preinstall script to wipe the `node_modules` folder that `npm install` just created.

I'm not sure what the original intent was here, so I can't recommend a fix.

**Comment by gkalpak** [comment]
I don't see how the changes on #12792 could cause this issue.
The `node_modules/` directory is purged before `npm install`, then `npm install` runs and creates the `node_modules/` directory (installing all necessary modules) and finally the (already existing) `npm-shrinkwrap.json` file is copied over to `node_modules/npm-shrinkwrap.cached.json` (which is created then; it doesn't exist already).

The only thing that would make sense, would be if somehow the underlying file created with [`var ws = fs.createWriteStream(dstPath);`][1] (where `dstPath` is `node_modules/npm-shrinkwrap.cached.json`) isn't created in time for getting written to (with [`rs.pipe(ws);`][2].
But this shouldn't be the case, since this is supposed to be taken care of (for quire some time now) in Node.js' [`fs` module][3].

--
I am clueless here :( @schmod, can you reproduce it consistently ?

<sub>I have only tried on npm v2.x, but I would be surprized if it tuned out to be npm-specific.</sub>

[1]: https://github.com/angular/angular.js/blob/80881949fcea9279b2eae195a70a993e8999b07e/scripts/npm/copy-npm-shrinkwrap.js#L35
[2]: https://github.com/angular/angular.js/blob/80881949fcea9279b2eae195a70a993e8999b07e/scripts/npm/copy-npm-shrinkwrap.js#L39
[3]: https://github.com/nodejs/node/blob/57359cd1e45084ac107ca5b16fe2a1423139d711/lib/fs.js#L1940

**Comment by schmod** [comment]
I just did a fresh clone of the repo on another computer with NPM 3.3.9, and duplicated the issue.

It appears that this *is* NPM-specific, as NPM's docs only seem to promise that a package's `preinstall` script will be executed before that same package's `install` script.  It specifically does not promise that the script will run before npm downloads and installs the package's dependencies.

This may be the same issue as https://github.com/npm/npm/issues/9916, although I suspect that very few packages are depending on the NPM lifecycle functioning this way.  

tl;dr; Angular's `preinstall` script appears to depend on an undocumented behavior in NPM that has changed in version 3.x.

**Comment by gkalpak** [comment]
I confirm that it doesn't work with npm v3.3.9. It seems that the `preinstall` scripts are run after the dependencies have been installed. I suggest we wait and see what is the response to npm/npm#9916.

@IgorMinar, any insight on this ? This should also affect ng2 (but I haven't tried it).

**Comment by schmod** [comment]
It looks like Node 5.0.0 will be shipping soon, and will be bundled with NPM 3.6.6.  

https://github.com/nodejs/node/pull/3466?utm_source=nodeweekly&utm_medium=email

**Comment by gkalpak** [comment]
This happens indeed with ng2 as well (node v4.2.1, npm v3.3.9).
Node v5.0.0 [has been released][1] with npm v3.3.6, so this should also affect users on node v5 on both ng1 and ng2 :(

[1]: https://nodejs.org/en/blog/release/v5.0.0/

**Comment by huguangju** [comment]
Whatever version, i didn't download npm package success :(

**Comment by gkalpak** [comment]
@huguangju, this issue specifically tracks issues with nm v3.x.
Your sounds like a different issue, so if you think this is somehow related to AngularJS (and not for example to node or npm), please open a new issue, providing relevant info (e.g. platform, node/npm version, error messages/faulty behaviour, steps to reproduce etc).

**Comment by wbrycki** [comment]
Is there a workaround to this issue?
Duplicated on npm 3.6.0

**Comment by gkalpak** [comment]
@wbrycki, you could try commenting out the `preinstall` script hook in `package.json`:

```json
  "scripts": {
    //"preinstall": "node scripts/npm/check-node-modules.js --purge",
    ...
```

That's causing one problem for sure. I'm not sure if there are other problems (e.g. dependencies incompatibility issues), but you can give it a shot.

**Comment by umerthow** [comment]
just restart you computer and try again to intalling npm by cmd

**Comment by Narretz** [comment]
We've since switched to node 6 and npm 3, so this should work now

**Comment by mgol** [comment]
To be precise, we're now on Node 6 and Yarn, not npm 3.
